File: 2025-11/2025-11-07/countingCards.js
function combinations(cards) {
  const cardsInDeck = 13 * 4;

  return factorial(cardsInDeck) / (factorial(cards) * factorial(cardsInDeck - cards));
}

const factorial = n => {
  if (n == 1 || n == 0)
    return 1;

  return n * factorial(n - 1);
}

console.log(combinations(52));
console.log(combinations(1));
console.log(combinations(2));
console.log(combinations(5));
console.log(combinations(10));
console.log(combinations(50));

console.log(combinations(0));

// --- TEST SUITE ---

const testsText = `
1. combinations(52) should return 1.
2. combinations(1) should return 52.
3. combinations(2) should return 1326.
4. combinations(5) should return 2598960.
5. combinations(10) should return 15820024220.
6. combinations(50) should return 1326.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a) {
    if (Array.isArray(a[index])) {
      if (Array.isArray(b[index])) {
        if (!arraysEqual(a[index], b[index]))
          return false;
      }
      else 
        return false;
    }

    else if (a[index] !== b[index])
      return false;
  }

  return true;
}

runTests(testData);

File: 11 November/11-07/counting_cards.py
def combinations(cards):
    cards_in_deck = 13 * 4

    return factorial(cards_in_deck) / (factorial(cards) * factorial(cards_in_deck - cards))

def factorial(n):
    if n == 1 or n == 0:
        return 1

    return n * factorial(n - 1)

print(combinations(52))
print(combinations(1))
print(combinations(2))
print(combinations(5))
print(combinations(10))
print(combinations(50))

print(combinations(0))

# --- TEST SUITE ---

tests_text = '''
1. combinations(52) should return 1.
2. combinations(1) should return 52.
3. combinations(2) should return 1326.
4. combinations(5) should return 2598960.
5. combinations(10) should return 15820024220.
6. combinations(50) should return 1326.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)